api-admin/src/main/java/bio/terra/pearl/api/admin/service/notifications/SendgridActivityService.java

```diff
@@ -0,0 +1,70 @@
+package bio.terra.pearl.api.admin.service.notifications;
+
+import bio.terra.pearl.core.dao.notification.SendgridEventDao;
+import bio.terra.pearl.core.model.notification.SendgridEvent;
+import bio.terra.pearl.core.service.notification.email.SendgridClient;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.List;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+@Service
+public class SendgridActivityService {
+  private static final Logger logger = LoggerFactory.getLogger(SendgridActivityService.class);
+  private SendgridClient sendgridClient;
+  private SendgridEventDao sendgridEventDao;
+
+  public SendgridActivityService(SendgridClient sendgridClient, SendgridEventDao sendgridEventDao) {
+    this.sendgridEventDao = sendgridEventDao;
+    this.sendgridClient = sendgridClient;
+  }
+
+  public void saveSendgridActivity() {
+    // Find the most recently recorded event, and get all events since then
+    Optional<SendgridEvent> mostRecentEvent = sendgridEventDao.findMostRecentEvent();
+
+    // If we've never recorded an event, default to 30 days ago. Sendgrid only keeps
+    // 30 days of activity history, so this will backfill as much history as possible.
+    Instant mostRecentEventDate =
+        mostRecentEvent
+            .map(SendgridEvent::getLastEventTime)
+            .orElse(Instant.now().minus(30, ChronoUnit.DAYS));
+
+    List<SendgridEvent> events = getAllRecentSendgridEvents(mostRecentEventDate, Instant.now());
+    // Over the course of gathering the results from Sendgrid's paginated API, it's possible that
+    // we've accumulated duplicate events at the fringes of the pages. The database will handle
+    // this fine, but this could result in the following log message indicating that we've inserted
+    // more events than we actually have.
+    logger.info("Sendgrid activity query returned {} events", events.size());
+
+    sendgridEventDao.bulkUpsert(events);
```

Comment:
this should be pulled out into a separate "bulkUpsert" method in the service class.  We want to make sure that updates to the database are wrapped in `@Transactional`, but we don't want to wrap this entire method, since we don't want to keep a transaction open while we're doing external API calls.